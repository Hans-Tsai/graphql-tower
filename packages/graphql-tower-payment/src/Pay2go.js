import url from 'url';
import crypto from 'crypto';
import assertResult from 'graphql-tower-helper/assertResult';

Date.now = jest.fn(() => 1517249431575);

export default class Pay2goPayment {
  constructor(env: Object = {}) {
    assertResult(env.PAY2GO_URL, new TypeError('PAY2GO_URL is required'));

    try {
      const uri = url.parse(env.PAY2GO_URL);
      const [hashIV, hashKey] = uri.auth.split(':');
      const host = `https://${uri.host}${uri.pathname}`;
      const merchantId = /id=(\w+)/i.exec(uri.query)[1];

      this.hashIV = hashIV;
      this.hashKey = hashKey;
      this.host = host;
      this.merchantId = merchantId;
      this.version = '1.2';
    } catch (e) {
      // empty
    }

    assertResult(
      this.hashIV && this.hashKey && this.host && this.merchantId,
      new TypeError('PAY2GO_URL was not in a correct format'),
    );
  }

  generateTemporaryCredentials(orderNo, amount) {
    const timeStamp = Date.now();
    const checkValue = [
      `HashKey=${this.hashKey}`,
      `Amt=${amount}`,
      `MerchantID=${this.merchantId}`,
      `MerchantOrderNo=${orderNo}`,
      `TimeStamp=${timeStamp}`,
      `Version=${this.version}`,
      `HashIV=${this.hashIV}`,
    ].join('&');

    return {
      merchantId: this.merchantId,
      merchantOrderNo: orderNo,
      timeStamp,
      amount,
      version: this.version,
      checkValue: crypto.createHash('sha256').update(checkValue, 'utf8').digest('hex'),
    };
  }
}
