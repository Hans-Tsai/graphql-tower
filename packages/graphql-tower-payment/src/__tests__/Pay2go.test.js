import Pay2go from '../Pay2go';

describe('pay2go', () => {
  it('is required', () => {
    expect(() => new Pay2go()).toThrowError(new TypeError('PAY2GO_URL is required'));
  });

  it('PAY2GO_URL was not in a correct format', () => {
    expect(() => new Pay2go({ PAY2GO_URL: 'https://host/name' }))
      .toThrowError(new TypeError('PAY2GO_URL was not in a correct format'));
  });

  it('successfully generate credentials', () => {
    const pay2go = new Pay2go({
      PAY2GO_URL: 'https://nApJ6NPsN95qN2cU:fuJsiQOHDJwpcUuffGFSJpY9fYrNII2U@ccore.spgateway.com/MPG/mpg_gateway?id=MS33234675',
    });

    expect(pay2go.generateTemporaryCredentials('XXX', 1000)).toMatchSnapshot();
  });
});
