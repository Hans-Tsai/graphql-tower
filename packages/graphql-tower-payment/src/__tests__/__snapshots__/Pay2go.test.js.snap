// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pay2go successfully generate credentials 1`] = `
Object {
  "amount": 1000,
  "checkValue": "980953c28ec39168b89790f3a42edfc2ceac23e5bcb0130fe30bdea13eafbc50",
  "merchantId": "MS33234675",
  "merchantOrderNo": "XXX",
  "timeStamp": 1517249431575,
  "version": "1.2",
}
`;
